icicle-cpu/src/lifter/mod.rs

```diff
@@ -73,17 +82,18 @@ impl InstructionLifter {
 
     /// Lift a single instruction starting at `vaddr` returning the address of the next instruction,
     /// or `None` if no instruction could be fetched from `vaddr`.
-    pub fn lift<S>(&mut self, src: &mut S, vaddr: u64) -> Option<u64>
-    where
-        S: InstructionSource,
+    pub fn lift<S>(&mut self, src: &mut S, vaddr: u64) -> Result<u64, DecodeError>
+        where
+            S: InstructionSource,
     {
         let next = self.decode(src, vaddr)?.inst_next;
         if self.generate_disassembly {
             self.disasm_current(src);
             tracing::trace!("disasm: {vaddr:#x} \"{}\"", self.disasm)
         }
 
-        let block = self.lifter.lift(&src.arch().sleigh, &self.decoded).ok()?;
+        let block = self.lifter.lift(&src.arch().sleigh, &self.decoded)
+            .map_err(|_| DecodeError::InvalidInstruction)?;
```

Comment:
BTW, If the lifter fails because it's unable to attach a token to a varnode (`UnknownVarNode`) should the disassembler backtrack and try the next constructor?

This is probably too much of an edge case that never happens, but having those errors separated could be helpful to find-out in the future.